src/__tests__/non-det-interpreter.ts

```diff
@@ -492,28 +505,62 @@ export async function testDeterministicCode(
 export async function testNonDeterministicCode(
   code: string,
   expectedValues: any[],
-  hasError: boolean = false
+  hasError: boolean = false,
+  random: boolean = false
 ) {
-  const context = makeNonDetContext()
+  const context: Context = makeNonDetContext()
   let result: Result = await runInContext(code, context, nonDetTestOptions)
 
+  const results: any[] = []
   const numOfRuns = hasError ? expectedValues.length - 1 : expectedValues.length
   for (let i = 0; i < numOfRuns; i++) {
-    expect((result as SuspendedNonDet).value).toEqual(expectedValues[i])
-    expect(result.status).toEqual('suspended-non-det')
+    if (random) {
+      results.push((result as SuspendedNonDet).value)
+    } else {
+      expect((result as SuspendedNonDet).value).toEqual(expectedValues[i])
+    }
 
+    expect(result.status).toEqual('suspended-non-det')
     result = await resume(result)
   }
 
-  if (!hasError) {
-    // all non deterministic programs have a final result whose value is undefined
-    expect(result.status).toEqual('finished')
-    expect((result as Finished).value).toEqual(undefined)
-  } else {
-    expect(result.status).toEqual('error')
-    const message: string = parseError(context.errors)
-    expect(message).toEqual(expectedValues[expectedValues.length - 1])
+  if (random) {
+    handleRandomizedTest(results, expectedValues)
+  }
+
+  if (hasError) {
+    handleErroneousTest(result, expectedValues, context)
```

Comment:
Let's name this method as `verifyError`, so the purpose of the method is more obvious.
The word `handle` may not be clear to new readers.